compiler/src/dotty/tools/dotc/transform/ElimJavaPackages.scala

```diff
@@ -0,0 +1,46 @@
+package dotty.tools.dotc.transform
+
+import dotty.tools.dotc.ast.tpd._
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.core.Types.{Type, TypeRef}
+import dotty.tools.dotc.transform.TreeTransforms.{MiniPhaseTransform, TransformerInfo}
+
+/**
+  * Eliminates syntactic references to Java packages, so that there's no chance
+  * they accidentally end up in the backend.
+  */
+class ElimJavaPackages extends MiniPhaseTransform {
+
+  override def phaseName: String = "elimJavaPackages"
+
+  override def transformSelect(tree: Select)(implicit ctx: Context, info: TransformerInfo): Tree = {
+    if (isJavaPackage(tree)) {
+      assert(tree.tpe.isInstanceOf[TypeRef], s"Expected tree with type TypeRef, but got ${tree.tpe.show}")
```

Comment:
How performance-sensitive is the miniphase code? Should I get rid of the assertion (and the asInstanceOf call below) in favour of *some* code duplication?